ignite/engine/engine.py

```diff
@@ -982,11 +991,15 @@ def _internal_run_as_gen(self) -> Generator[Any, None, State]:
                     # time is available for handlers but must be updated after fire
                     self.state.times[Events.EPOCH_COMPLETED.name] = epoch_time_taken
 
-                    handlers_start_time = time.time()
-                    self._fire_event(Events.EPOCH_COMPLETED)
-                    epoch_time_taken += time.time() - handlers_start_time
-                    # update time wrt handlers
-                    self.state.times[Events.EPOCH_COMPLETED.name] = epoch_time_taken
+                    if not self._skip_epoch_completed_after_termination:
+                        handlers_start_time = time.time()
+                        self._fire_event(Events.EPOCH_COMPLETED)
+                        epoch_time_taken += time.time() - handlers_start_time
+                        # update time wrt handlers
+                        self.state.times[Events.EPOCH_COMPLETED.name] = epoch_time_taken
+                    else:
+                        self._skip_epoch_completed_after_termination = False
```

Comment:
I'm not totally certain about this else part. Should we really have this code?

Reply:
I see, you are right. Thinking more about all these new flags and `should_terminate` and `should_terminate_single_epoch`. Maybe, instead of introducing new ones, we could enlarge the field of possible value of existing `should_terminate_single_epoch` or `should_terminate`:
Now, they are bools:
```python
should_terminate_single_epoch: bool  # True, False
should_terminate: bool  # True, False
```
and they can be 
```python
should_terminate_single_epoch: bool | str  # True, False, "skip_epoch_completed"
should_terminate: bool | str  # True, False, "skip_completed"
```
This way we wont need to add extra resets for the new flags. What do you think?

My another concern now also is about https://github.com/pytorch/ignite/blob/e2980deb7c8e8164feda3b084a2c69141799524c/ignite/engine/engine.py#L971
where we reset the flags in the very beginning. It is possible that the new flags should be also reset (as you suggested probably in the previous PR) and this can be visible on some multi-run tests as the new flags have the state now...

Reply:
I agree that four flags are probably not necessary. How about making `should_terminate` and `should_terminate_single_epoch ` integers? This would preserve the compatibility with third-party code involving bool checks on those variables.

Reply:
> This would preserve the compatibility with third-party code involving bool checks on those variables.

Good point, how it would work with integers you think about ?

Reply:
@vfdev-5 I still have to think about it, but maybe we could define both `should_terminate` and `should_terminate_single_epoch` as integers where:

- 0 means that no termination has been requested
- 1 means that a termination has been requested
- 2 means that a termination has been requested with the flag `skip_complete=True`.

Do you think this would improve the readability of the code, though?

Reply:
I was thinking about your first point:
> This would preserve the compatibility with third-party code involving bool checks on those variables.

it means that let's say some 3rd party code does `isinstance(engine.should_terminate, bool)` and in case of the new value when we do `engine.should_terminate == "skip_completed"`, their check `isinstance(engine.should_terminate, bool) == False` would not work. 

On the other, with integer type of `engine.should_terminate` the check wont pass even with old values, so this will be BC-breaking.

I think it is fine to keep adopt:
```
should_terminate_single_epoch: bool | str  # True, False, "skip_epoch_completed"
should_terminate: bool | str  # True, False, "skip_completed"
```
What do you think ?

Reply:
Thanks for working on this PR, sounds good to me!

Reply:
Done in commit [cca9beb](https://github.com/bonassifabio/ignite/commit/cca9bebf224106e44e202175fb896579d08c741c)